packages/@uppy/aws-s3-multipart/src/MultipartUploader.js

```diff
@@ -219,32 +238,46 @@ class MultipartUploader {
     })
   }
 
-  _uploadPartRetryable (index) {
+  _prepareUploadParts (candidates) {
```

Comment:
Can we use a private method here? Given https://github.com/transloadit/uppy/pull/3076, it would make sense to use that instead of a trailing underscore method I think. Also I'd prefer to make it async to avoid `Promise.resolve().then` pattern that we don't need now that we've dropped IE support.
```suggestion
  async #prepareUploadParts (candidates) {
```

Reply:
Sorry @aduh95 can you please elaborate on the `Promise.resolve().then` pattern, I am unfamiliar with its use. For example, for `prepareUploadParts` the change seems fairly straightforward:

```javascript
async #prepareUploadParts (candidates) {
  this.lockedCandidatesForBatch.push(...candidates)
  const result = await this.options.prepareUploadParts({
    key: this.key,
    uploadId: this.uploadId,
    partNumbers: candidates.map((index) => index + 1),
  })
  const valid = typeof result?.presignedUrls === 'object'
  if (!valid) {
    throw new TypeError(
      'AwsS3/Multipart: Got incorrect result from `prepareUploadParts()`, expected an object `{ presignedUrls }`.'
    )
  }

  return result
}
```

This works just fine, but if I change `#uploadPart` in a similar way the tests just time out and never resolve. But I am not even sure if I'm changing the right thing here, because I don't know the reason for returning `Promise.resolve().then()` in the first place instead of just returning the actual promise:

```javascript
async #uploadPart (index, prePreparedPart) {
  const body = this.chunks[index]
  this.chunkState[index].busy = true

  const valid = typeof prePreparedPart?.url === 'string'
  if (!valid) {
    throw new TypeError('AwsS3/Multipart: Got incorrect result for `prePreparedPart`, expected an object `{ url }`.')
  }

  const ({url, headers} = prePreparedPart)
  if (this.#aborted()) {
    this.chunkState[index].busy = false
    throw createAbortError()
  }

  return this.#uploadPartBytes(index, url, headers)
}
```

Then in `#completeUpload`  and `#createUpload` things get really strange; why not just `return this.options.completeMultipartUpload({` in this case? What is the alternative, and is error handling for async/await just done with `try/catch` blocks instead?

```javascript
#completeUpload () {
  // Parts may not have completed uploading in sorted order, if limit > 1.
  this.parts.sort((a, b) => a.PartNumber - b.PartNumber)

  return Promise.resolve().then(() => this.options.completeMultipartUpload({
    key: this.key,
    uploadId: this.uploadId,
    parts: this.parts,
  })).then((result) => {
    this.options.onSuccess(result)
  }, (err) => {
    this.#onError(err)
  })
}
```

Can you please give me a hand with this

I've replaced the underscores with `#` for the private methods just fine.

Reply:
`Promise.resolve().then(` is useful when:
- we want to queue a microstask in an environment where `queueMicrotask` exists.
- want to turn thrown error into a promise rejection in an environment where `async`/`await` is not available.

I don't think there's any reason to use it anymore given we can now use `async`/`await` and `queueMicrotask`, we haven't gotten rid of it just yet.

> if I change `#uploadPart` in a similar way the tests just time out and never resolve

Maybe you comment is lacking some context, but the code you pasted changes the function signature (it now expects two parameter instead of one), I would assume that's why the test do not complete anymore.

> why not just `return this.options.completeMultipartUpload({` in this case? What is the alternative, and is error handling for async/await just done with try/catch blocks instead?

You're right, this code could be indeed simplified using `try { return await this.options.completeMultipartUpload(...`.


> I've replaced the underscores with `#` for the private methods just fine.

Thanks for doing that, I plan on landing https://github.com/transloadit/uppy/pull/3076 which does effectively that, I'll resolve conflicts if any arises.

Reply:
@aduh95 thank you for that, I just didn't know what to search for and your comments about `queueMicrotask` was helpful in finding the right things. I've changed all I can to use async/await now, the only part I left was in `createUpload` because that is a special case used for cancelling uploads:

```javascript
#createUpload () {
  this.createdPromise = Promise.resolve().then(() => this.options.createMultipartUpload())
  return this.createdPromise.then((result) => {
```

Reply:
And I am not sure what I was doing wrong with `#uploadPart` yesterday, must have just been a syntax error! :sweat: